vtex/go-io
# Fix Redis connection concurrent usage

#### What is the purpose of this pull request?
Earlier this week we got an error on Colossus where a specific instance of
Colossus started getting a "closed connection" error on every subscription
requested by clients. This is the query on Splunk to see how strange that
behavior was: [[Splunk]](https://splunk7.vtex.com/en-US/app/vtex_kuberouter/search?q=search%20index%3Dcolossus%20(code%20IN%20(pubsub_error%20subscription_error%20redis_conn_error%20redis_conn_reset_error))%20%0A%7C%20eval%20count_bad_host%3Dif(host%3D%22colossus0-39-2-dev-66f589bb57-c9t8t%22%2C%201%2C%20null)%0A%7C%20eval%20count_other_hosts%3Dif(host%3D%22colossus0-39-2-dev-66f589bb57-c9t8t%22%2C%20null%2C%201)%0A%7C%20timechart%20sum(count_bad_host)%20sum(count_other_hosts)%20by%20code&display.page.search.mode=fast&dispatch.sample_ratio=1&workload_pool=&earliest=1594634400&latest=1594666800&display.general.type=visualizations&display.page.search.tab=visualizations&display.visualizations.trellis.enabled=1&display.visualizations.charting.legend.placement=right&display.visualizations.charting.axisTitleX.visibility=visible&display.visualizations.charting.axisTitleY.visibility=visible&display.visualizations.charting.axisTitleY2.visibility=visible&display.visualizations.trellis.splitBy=_aggregation&display.visualizations.trellis.size=large&display.visualizations.charting.chart=line&display.visualizations.charting.axisY.scale=log&sid=1594938269.735207_B1AFCE35-A4A4-40A6-92E9-145382F27731)

I've looked around the code, and couldn't find any reason how we could
ever get into that weird state, but investigating deeper into the Redis library
that we use, I found that it is actually not thread-safe for concurrent
reads or writes under the same request: [[StackOverflow]](https://stackoverflow.com/questions/37178897/some-questions-about-redigo-and-concurrency) :harold-pain:

So in this PR, apart from moving stuff around to improve overall quality,
I've decided to:
 - Move all commands performed in the redis connection to the single
background loop, thus avoiding any concurrent access to the same
connection;
 - Improve the logs, so that we can have more information in case this
problem ever happens again in the new version;
 - _TODO:_ We should have alarms to know ASAP if the problem happens.

This actually also improves our resiliency, as the request for a command will
be simply sending to a channel which will never fail, while we can afford 
indefinitely retrying the request in the background "main loop". I've made sure
to do only 1, explicit, retry of `PSubscribe` command though, since I'm not
sure if we may end up calling this function with a bad subscription that
always fails when we send to Redis.
(The alternative would be to always add the requested patterns to that
`subscribedPatterns` field, so that if the individual command fails, we'd
retry indefinitely on the "recover connection loop")

#### What problem is this solving?
Concurrent access to a **non** thread-unsafe shared resource.

#### How should this be manually tested?
 - Check the new logs are now sent to Splunk

Stretch (?): 
 - Add some "chaos monkey" code to randomly return an error sometimes
 - Ensure connection recovery works as expected (through those same logs)

#### Types of changes
